### PR TITLE
WIP: Add Labels to Milestones

### LINKED ISSUE #17244: [Feature Request] Milestone Labels
### Feature Description

I think it would be super helpful for milestones (and the associated APIs) to support labels. They help quite a bit organizing issues, and on large projects it would be nice to be able to categorize the milestones as well. 

Right now in our projects we have to resort to using special keywords in the titles which isn't super-efficient or clear like you see with the labels in issues.

### Screenshots

**1.** Labels could appear below the description in the milestone listing.

Screenshot

**2.** Leverage the same type of dropdown when creating/editing, just show labels in a horizontal stack.
Screenshot2
)

**3.** Bonus ask: Would probably also be good to *see* (likely not edit) the milestone labels on any assigned issue.

Screenshot3
.

### PR DESCRIPTION
**Pull request by techknowlogick**
Fix #17244

Marked as WIP as I haven't completed the UI yet, but the API and the structs/migrations are complete.

cc: @chriseaton

### CODE REVIEW

File: models/migrations/v197.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import (
+	"fmt"
+
+	"xorm.io/xorm"
+)
+
+func addLabelsToMilestones(x *xorm.Engine) error {
+	type MilestoneLabel struct {
+		ID          int64 `xorm:"pk autoincr"`
+		MilestoneID int64 `xorm:"UNIQUE(s)"`
```
Comment (delvh) [inline models/migrations/v197.go]: What do you mean by `UNIQUE(s)`?
Shouldn't it be `UNIQUE(milestoneid_labelid)` or something similar?

File: models/milestone_label.go [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+	}
+
+	sess := db.NewSession(db.DefaultContext)
+	defer sess.Close()
+	if err = sess.Begin(); err != nil {
+		return err
+	}
+
+	// Do NOT add invalid labels
+	if milestone.RepoID != label.RepoID && milestone.Repo.OwnerID != label.OrgID {
```
Comment (delvh) [inline models/milestone_label.go:74]: I can understand the first condition, but why the second condition?
And why is it combined with an `&&`?

File: models/milestone_label.go [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+	return sess.Commit()
+}
+
+// newMilestoneLabels add labels to an milestone. It will check if the labels are valid for the milestone
+func newMilestoneLabels(e db.Engine, milestone *Milestone, labels []*Label, doer *User) (err error) {
+	for _, label := range labels {
+		// Don't add already present labels and invalid labels
+		if hasMilestoneLabel(e, milestone.ID, label.ID) ||
+			(label.RepoID != milestone.RepoID && label.OrgID != milestone.Repo.OwnerID) {
```
Comment (delvh) [inline models/milestone_label.go:95]: I have the feeling that this check should be extracted into its own function as you use it more than once.
And I'm still confused by it.
It could be called i.e. `isMilestoneLabelPairValid` or something like that.

File: models/milestone_label.go [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+	return nil
+}
+
+// NewMilestoneLabels creates a list of milestone-label relations.
+func NewMilestoneLabels(milestone *Milestone, labels []*Label, doer *User) (err error) {
+	ctx, committer, err := db.TxContext()
+	if err != nil {
+		return err
+	}
+	defer committer.Close()
```
Comment (delvh) [inline models/milestone_label.go:113]: Shouldn't the `ctx` be closed as well, or does the committer suffice?

File: models/milestone_label.go [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+	}
+
+	if err = deleteMilestoneLabel(sess, milestone, label, doer); err != nil {
+		return err
+	}
+
+	milestone.Labels = nil
+	if err = milestone.loadLabels(sess); err != nil {
+		return err
+	}
```
Comment (delvh) [inline models/milestone_label.go]: Why are those lines necessary?
That goes far beyond deleting the relation and rather seems like a copy-paste error, doesn't it?
```suggestion
```

File: models/milestone_label.go [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+	return nil
+}
+
+func (milestone *Milestone) removeLabel(e db.Engine, doer *User, label *Label) error {
+	return deleteMilestoneLabel(e, milestone, label, doer)
+}
+
+func (milestone *Milestone) clearLabels(e db.Engine, doer *User) (err error) {
+	if err = milestone.getLabels(e); err != nil {
```
Comment (delvh) [inline models/milestone_label.go]:
```suggestion
	if err = milestone.loadLabels(e); err != nil {
```

File: models/milestone_label.go [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+}
+
+// ClearLabels removes all milestone labels as the given user.
+// Triggers appropriate WebHooks, if any.
+func (milestone *Milestone) ClearLabels(doer *User) (err error) {
+	ctx, committer, err := db.TxContext()
+	if err != nil {
+		return err
+	}
+	defer committer.Close()
```
Comment (delvh) [inline models/milestone_label.go:216]: See above regarding `ctx`.

File: models/milestone_label.go [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+	}
+
+	if err = committer.Commit(); err != nil {
+		return fmt.Errorf("Commit: %v", err)
+	}
+
+	return nil
+}
+
+// ReplaceLabels removes all current labels and add new labels to the milestone.
```
Comment (delvh) [inline models/milestone_label.go]:
```suggestion
// ReplaceLabels removes all current labels and adds the given labels to the milestone.
```

File: models/milestone_label.go [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+	return sess.Commit()
+}
+
+// LoadLabels loads labels
+func (milestone *Milestone) LoadLabels() error {
+	return milestone.loadLabels(db.GetEngine(db.DefaultContext))
+}
+
+func (milestone *Milestone) loadLabels(e db.Engine) (err error) {
+	if milestone.Labels == nil {
```
Comment (delvh) [inline models/milestone_label.go:160]: Only this line differs between this method and `getLabels`:
`if len(milestone.Labels) > 0 {`
Which one should be taken then?
Afterwards, the other method can be deleted.

File: models/milestone_label.go [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+	if len(milestone.Labels) > 0 {
+		return nil
+	}
+
+	milestone.Labels, err = getLabelsByMilestoneID(e, milestone.ID)
+	if err != nil {
+		return fmt.Errorf("getLabelsByMilestoneID: %v", err)
+	}
+	return nil
+}
```
Comment (delvh) [inline models/milestone_label.go]:
```suggestion
```

File: models/milestone_label.go [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+	for _, l := range toRemove {
+		if err = milestone.removeLabel(db.GetEngine(ctx), doer, l); err != nil {
+			return fmt.Errorf("removeLabel: %v", err)
+		}
+	}
+
+	milestone.Labels = nil
+	if err = milestone.loadLabels(db.GetEngine(ctx)); err != nil {
+		return err
+	}
```
Comment (delvh) [inline models/milestone_label.go:296]: That seems overly complicated for what you are trying to achieve.
I thought you wanted to clear all labels and set only the ones given as a parameter?
That can be achieved much more clearly: Drop all labels and then add all given labels to the milestone.
I refuse to review this method in the current implementation.

File: routers/api/v1/repo/milestone_label.go [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+	}
+
+	ctx.Status(http.StatusNoContent)
+}
+
+// ReplaceMilestoneLabels replace labels for a milestone
+func ReplaceMilestoneLabels(ctx *context.APIContext) {
+	// swagger:operation PUT /repos/{owner}/{repo}/milestones/{id}/labels milestone milestoneReplaceLabels
+	// ---
+	// summary: Replace a milestone's labels
```
Comment (delvh) [inline routers/api/v1/repo/milestone_label.go]:
```suggestion
	// summary: Drop all previous milestone labels and replace them with new labels
```

File: routers/api/v1/repo/milestone_label.go [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+	//   description: id of the label to remove
+	//   type: integer
+	//   format: int64
+	//   required: true
+	// responses:
+	//   "204":
+	//     "$ref": "#/responses/empty"
+	//   "403":
+	//     "$ref": "#/responses/forbidden"
+	//   "422":
```
Comment (delvh) [inline routers/api/v1/repo/milestone_label.go:146]: Where does it return that?
The `http.StatusUnprocessableEntity`?

File: models/issue_milestone.go [hunk 14]
```diff
@@ -97,6 +98,26 @@ func NewMilestone(m *Milestone) (err error) {
...
+		// During the session, SQLite3 driver cannot handle retrieve objects after update something.
+		// So we have to get all needed labels first.
+		labels := m.Labels
+		if err = sess.Find(&m.Labels); err != nil {
+			return fmt.Errorf("find all labels [label_ids: %v]: %v", m.Labels, err)
+		}
+
+		for _, label := range labels {
+			// Silently drop invalid labels.
+			if label.RepoID != m.Repo.ID && label.OrgID != m.Repo.OwnerID {
```
Comment (delvh) [inline models/issue_milestone.go:111]: As mentioned: Make that a method! I've seen it at list four times now.

File: routers/web/repo/milestone.go [hunk 15]
```diff
@@ -116,6 +117,41 @@ func NewMilestone(ctx *context.Context) {
 	ctx.Data["Title"] = ctx.Tr("repo.milestones.new")
 	ctx.Data["PageIsIssueList"] = true
 	ctx.Data["PageIsMilestones"] = true
+	var labelIDs []int64
+	var err error
+	selectLabels := ctx.FormString("labels")
+	if len(selectLabels) > 0 && selectLabels != "0" {
```
Comment (delvh) [inline routers/web/repo/milestone.go:123]: Why can't it be `0`? Is such a label name prohibited?

File: templates/repo/issue/milestone_labels/label.tmpl [hunk 16]
```diff
@@ -0,0 +1,9 @@
+<a
```
Comment (delvh) [inline templates/repo/issue/milestone_labels/label.tmpl:1]: Is there a reason why you don't just use the template that is used for issue labels?
Because if there is not, then I'd recommend for maintainability purposes to use only the existing one.

File: templates/repo/issue/milestone_labels/labels_sidebar.tmpl [hunk 17]
```diff
@@ -0,0 +1,11 @@
+<div class="ui labels list">
```
Comment (delvh) [inline templates/repo/issue/milestone_labels/labels_sidebar.tmpl:1]: Again, isn't there an `issue` equivalent you can use?
If yes, then please use that to improve future maintainability.

File: web_src/js/index.js [hunk 18]
```diff
@@ -393,8 +393,8 @@ function initSimpleMDEImagePaste(simplemde, dropzone, files) {
 let autoSimpleMDE;
 
 function initCommentForm() {
-  if ($('.comment.form').length === 0) {
-    return;
+  if ($('.comment.form').length === 0 && $('.select-label').length < 1) {
```
Comment (delvh) [inline web_src/js/index.js:399]: ?

File: templates/repo/issue/milestone_new.tmpl [hunk 19]
```diff
@@ -39,6 +39,40 @@
 					<label>{{.i18n.Tr "repo.milestones.desc"}}</label>
 					<textarea name="content">{{.content}}</textarea>
 				</div>
+				<div class="ui segment metas">
+					<input id="label_ids" name="label_ids" type="hidden" value="{{.label_ids}}">
```
Comment (delvh) [inline templates/repo/issue/milestone_new.tmpl:43]: Can't you reuse the template that is used for issues here as well? (If it isn't a template yet **make it**!)
Again, I don't see a reason why we should impact future maintainability if not necessary.

File: models/milestone_label.go (hunk 2 above)
Comment (techknowlogick) [reply models/milestone_label.go:74]: Because org labels should be allow too, so if the label doesn't belong to the repo OR the org THEN skip adding

**Comment by MrHBS** [comment]
@techknowlogick Are you still working on this?

**Comment by bqv** [comment]
@MrHBS are you?
